### PR TITLE
Feature/instructor types

### PR DESCRIPTION
**Pull request by ltwheeler**
Migrates old userRoles and budget entities to a more holistic approach to instructor types across IPA.

Will allow more granularity in the budget view, as well as assigning general types instead of people.

### CODE REVIEW

File: app/assignment/directives/courseAssignmentTable.js [hunk 1]
```diff
@@ -316,13 +332,32 @@ assignmentApp.directive("courseAssignmentTable", this.courseAssignmentTable = fu
...
+							instructorTypeId: instructorTypeId
+						};
+
+						// Remove The Staff if necessary
+						if (sectionGroup) {
+							sectionGroup.showTheStaff = false;
+						}
 
+						assignmentActionCreators.assignInstructorType(newTeachingAssignment);
 					// Create a 'The Staff' placeholder
```
Comment (cthielen) [inline app/assignment/directives/courseAssignmentTable.js]: Comment on code should go inside the case

File: app/assignment/services/assignmentActionCreators.js [hunk 2]
```diff
@@ -206,6 +174,35 @@ assignmentApp.service('assignmentActionCreators', function (assignmentStateServi
...
+			assignmentService.addInstructorAssignment(teachingAssignment, scheduleId).then(function (newTeachingAssignment) {
+				$rootScope.$emit('toast', { message: "Assigned instructor type", type: "SUCCESS" });
+				assignmentStateService.reduce({
+					type: ADD_TEACHING_ASSIGNMENT,
+					payload: {
+						teachingAssignment: newTeachingAssignment
+					}
+				});
+			}, function (err) {
+				$rootScope.$emit('toast', { message: "Could not assign instructor to course.", type: "ERROR" });
```
Comment (cthielen) [inline app/assignment/services/assignmentActionCreators.js]: Success is 'assigned instructor type' but failure is 'could not assign instructor to course'? Aren't those slightly different?

File: app/budget/budgetCtrl.js [hunk 3]
```diff
@@ -5,6 +5,8 @@ budgetApp.controller('BudgetCtrl', ['$scope', '$rootScope', '$window', '$locatio
 
 		$scope.view = {};
 
+		$scope.budgetConfigStyles = { "width" : "40%" };
```
Comment (cthielen) [inline app/budget/budgetCtrl.js]: Is there a reason this isn't in CSS?

File: app/budget/directives/courseCosts/courseCosts.html [hunk 4]
```diff
@@ -6,6 +6,18 @@
 		slim="true">
 
 		<table class="course-costs__table">
+			<colgroup>
+				<col span="1" style="width: 6%;">
```
Comment (cthielen) [inline app/budget/directives/courseCosts/courseCosts.html]: Isn't span="1" the default?

File: app/budget/directives/instructorCosts/instructorCosts.html [hunk 5]
```diff
@@ -5,16 +5,25 @@
 		select-tab="setActiveTerm"
 		slim="true">
 		<table class="instructor-costs__table">
+			<colgroup>
+				<col span="1" style="width: 10%;">
```
Comment (cthielen) [inline app/budget/directives/instructorCosts/instructorCosts.html]: Isn't span="1" the default?

File: app/budget/directives/lineItems/lineItemDropdown/lineItemDropdown.html [hunk 6]
```diff
@@ -18,6 +18,7 @@
 		<li class="line-item-row">
 			<a class="budget-scenario-item"
 			   uib-tooltip="Delete line item"
+				 tooltip-append-to-body="true"
```
Comment (cthielen) [inline app/budget/directives/lineItems/lineItemDropdown/lineItemDropdown.html]: Correct this indentation

File: app/budget/directives/modals/budgetConfig/groupCostConfig/groupCostConfig.css [hunk 7]
```diff
@@ -1,37 +1,26 @@
...
 }
 
-.group-cost-config__validation-error {
-	margin-bottom: 7px;
-	height: 17px;
-	color: #f26868;
+.group-cost__cost-container {
+	padding-top: 5px;
+	padding-bottom: 5px;
 }
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/groupCostConfig/groupCostConfig.css]: Add a newline at the end of the file

File: app/budget/directives/modals/budgetConfig/groupCostConfig/groupCostConfig.html [hunk 8]
```diff
@@ -1,79 +1,54 @@
...
-			</ipa-input>
-		</div>
-		<div class="group-cost-config__label">
-			TA Cost
-		</div>
-	</div>
+<div class="instructor-cost-config">
+	<table class="instructor-cost-config__table">
+		<colgroup>
+			<col span="1" style="width: 50%;">
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/groupCostConfig/groupCostConfig.html]: Isn't span="1" and half width (in the case of two 'cols') the defaults?

File: app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.css [hunk 9]
```diff
@@ -2,24 +2,24 @@
...
+.instructor-cost-config__cell {
+	padding: 10px;
 }
 
-.instructor-cost-config__type-container {
-	padding-left: 10px;
+.instructor-costs__cost-container {
+	padding-top: 5px;
+	padding-bottom: 5px;
 }
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.css]: Add a newline at the end of the file

File: app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.html [hunk 10]
```diff
@@ -1,22 +1,32 @@
...
-
-			<div class="instructor-cost-config__label">
-				{{ instructor.lastName }}, {{ instructor.firstName }}
-			</div>
-
-		</div>
-	</div>
+	<table class="instructor-cost-config__table">
+		<colgroup>
+			<col span="1" style="width: 33%;">
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.html]: Isn't span="1" and width="33%" the default (when there are three 'cols')?

File: app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.html [hunk 11]
```diff
@@ -1,22 +1,32 @@
...
+			<tr>
+				<td class="instructor-cost-config__cell">
+					{{ instructor.lastName }}, {{ instructor.firstName }}
+				</td>
+				<td class="instructor-cost-config__cell">
+					{{ instructor.instructorType.description }}
+				</td>
+				<td class="instructor-cost-config__cell">
+					<ipa-input mode="'currency'"
+						value="instructor.instructorCost.cost"
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.html]: Please line up the attributes when a tag spans multiple lines

File: app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.html [hunk 12]
```diff
@@ -1,22 +1,32 @@
...
+						on-update="updateInstructorCost(instructor.instructorCost)">
+					</ipa-input>
+				</td>
+			</tr>
+		</tbody>
+	</table>
 </div>
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/instructorCostConfig/instructorCostConfig.html]: Add a newline at the end of the file

File: app/budget/services/actions/budgetCalculations.js [hunk 13]
```diff
@@ -427,55 +500,164 @@ budgetApp.service('budgetCalculations', function ($rootScope, $window, budgetSer
...
-			var calculatedInstructors = [];
+			instructor.instructorCost = null;
+			instructor.description = instructor.lastName + ", " + instructor.firstName;
 
-			instructors.ids.forEach(function(instructorId) {
-				var instructor = instructors.list[instructorId];
-				instructor.instructorCost = null;
-				calculatedInstructors.push(instructor);
+			instructor.instructorType = this._calculateInstructorType(instructor.id);
+			// Attach instructorCost
```
Comment (cthielen) [inline app/budget/services/actions/budgetCalculations.js]: Insert newline above this comment

File: app/budget/services/actions/budgetCalculations.js [hunk 14]
```diff
@@ -427,55 +500,164 @@ budgetApp.service('budgetCalculations', function ($rootScope, $window, budgetSer
...
+			var instructorTypes = budgetReducers._state.instructorTypes;
+
+			var instructor = assignedInstructors.list[instructorId] || activeInstructors.list[instructorId];
+			var user = users.byLoginId[instructor.loginId.toLowerCase()];
+
+			if (!user) { return; }
+
+			if (userRoles.byUserId[user.id]) {
+				userRoles.byUserId[user.id].forEach(function(userRole) {
+					// 15 is the roleId of 'instructor'
```
Comment (cthielen) [inline app/budget/services/actions/budgetCalculations.js]: Define magic numbers as constants

File: app/scheduling/templates/SchedulingCtrl.html [hunk 15]
```diff
@@ -155,11 +155,15 @@ <h3>Locations</h3>
 								<!-- TODO: Display enrollment and total seats (Involves modifying payload to include all sections) -->
 
 								<div ng-if="view.state.sectionGroups.list[sectionGroupId].instructorIds.length">
-									instructor{{ view.state.sectionGroups.list[sectionGroupId].instructorIds.length > 1 ? 's':'' }}:
+									Instructor{{ view.state.sectionGroups.list[sectionGroupId].instructorIds.length > 1 ? 's':'' }}:
```
Comment (cthielen) [inline app/scheduling/templates/SchedulingCtrl.html]: Surround the : with spaces

File: app/scheduling/templates/SchedulingCtrl.html [hunk 16]
```diff
@@ -155,11 +155,15 @@ <h3>Locations</h3>
...
 									<div class="label label-default" style="padding: 3px; margin-left: 3px;"
 										ng-repeat="instructorId in view.state.sectionGroups.list[sectionGroupId].instructorIds track by $index">
 										{{ view.state.instructors.list[instructorId] | lastSpaceInitial }}{{ $last ? '':', '}}
 									</div>
+									<span class="label label-default" style="padding: 3px; margin-left: 3px;"
```
Comment (cthielen) [inline app/scheduling/templates/SchedulingCtrl.html]: Should this inline style be in a CSS file?

File: app/supportAssignment/directives/supportAssignmentSearch/supportAssignmentSearch.css [hunk 17]
```diff
@@ -4,4 +4,8 @@
 	align-items: center;
 	width: 40%;
 	margin-left: 20%;
+}
+
+.support-assignment-search__input {
+	width: 100%;
 }
```
Comment (cthielen) [inline app/supportAssignment/directives/supportAssignmentSearch/supportAssignmentSearch.css]: Add a newline at the end of the file

File: app/teachingCall/css/teaching-call-modal.css [hunk 18]
```diff
@@ -134,4 +134,10 @@
 
 .instructor-list {
 	height: 235px;
+}
+
+.instructor-toggle-btn-container .add-instructor-modal__instructor-btn {
+	margin-right: 3px;
+	margin-bottom: 3px;
+	width: fit-content;
 }
```
Comment (cthielen) [inline app/teachingCall/css/teaching-call-modal.css]: Add a newline at the end of the file

File: app/teachingCall/css/teaching-call-status.css [hunk 19]
```diff
@@ -209,4 +210,12 @@
...
 	padding-top: 20px;
+}
+
+.teaching-call-status__no-calls-splash {
+	display: flex;
+	justify-content: center;
+	margin-top: 100px;
+	font-size: 19px;
+	color: gray;
 }
```
Comment (cthielen) [inline app/teachingCall/css/teaching-call-status.css]: Add a newline at the end of the file

File: app/teachingCall/teachingCallApp.js [hunk 20]
```diff
@@ -34,3 +34,8 @@ var CONTACT_INSTRUCTORS = "CONTACT_INSTRUCTORS";
 var ADD_INSTRUCTORS_TO_TEACHING_CALL = "ADD_INSTRUCTORS_TO_TEACHING_CALL";
 var REMOVE_INSTRUCTOR_FROM_TEACHING_CALL = "REMOVE_INSTRUCTOR_FROM_TEACHING_CALL";
 var CHANGE_TERM = "CHANGE_TERM";
+
+// Calculations
+var CALCULATE_INSTRUCTORS_IN_CALL = "CALCULATE_INSTRUCTORS_IN_CALL";
+var CALCULATE_ELIGIBLE_INSTRUCTORS = "CALCULATE_ELIGIBLE_INSTRUCTORS";
+var SELECT_INSTRUCTORS = "SELECT_INSTRUCTORS";
```
Comment (cthielen) [inline app/teachingCall/teachingCallApp.js]: Add a newline at the end of the file

File: app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js [hunk 21]
```diff
@@ -30,18 +31,49 @@ teachingCallApp.controller('ModalAddInstructorsCtrl', ['$scope', '$rootScope', '
...
+	$scope.startTeachingCallConfig.invitedInstructors = state.calculations.instructorsEligibleForCall;
 	$scope.startTeachingCallConfig.invitedInstructors = _array_sortByProperty($scope.startTeachingCallConfig.invitedInstructors, "lastName");
 
-	$scope.startTeachingCallConfig.invitedInstructors.forEach(function(slotInstructor) {
-		slotInstructor.invited = false;
+	$scope.startTeachingCallConfig.invitedInstructors.forEach(function(instructor) {
+		instructor.invited = false;
 	});
 
+	$scope.areThereInstructorsOfType = function(instructorTypeId) {
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js]: Please rename this method. Perhaps instructorTypeUsed() ?

File: app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js [hunk 22]
```diff
@@ -30,18 +31,49 @@ teachingCallApp.controller('ModalAddInstructorsCtrl', ['$scope', '$rootScope', '
...
+		$scope.startTeachingCallConfig.invitedInstructors.forEach(function(instructor) {
+			if(instructor.instructorTypeId == instructorTypeId) {
+				instructor.invited = true;
+			}
+		});
+
+		$scope.startTeachingCallConfig.isAddInstructorFormComplete = $scope.isAddInstructorFormComplete();
+	};
+
+	$scope.areAllInstructorsOfTypeInvited = function (instructorTypeId) {
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js]: Please rename this method.

File: app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js [hunk 23]
```diff
@@ -30,18 +31,49 @@ teachingCallApp.controller('ModalAddInstructorsCtrl', ['$scope', '$rootScope', '
...
+		var areAllInvited = true;
+
+		$scope.startTeachingCallConfig.invitedInstructors.forEach(function(instructor) {
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js]: Use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find and get rid of the temporary variable

File: app/teachingCall/teachingCallStatus/services/teachingCallStatusActionCreators.js [hunk 24]
```diff
@@ -77,9 +89,132 @@ teachingCallApp.service('teachingCallStatusActionCreators', function (teachingCa
...
+				self._calculateEligibleInstructors();
 			}, function (err) {
 				$rootScope.$emit('toast', { message: "Could not remove instructor from teaching call.", type: "ERROR" });
 			});
+		},
+		toggleInstructor: function(instructorId) {
+			var selectedInstructorIds = teachingCallStatusStateService._state.ui.selectedInstructorIds;
+			var index = selectedInstructorIds.indexOf(instructorId);
+
+			if (index == -1) {
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/services/teachingCallStatusActionCreators.js]: Add a comment on what this is doing

File: app/teachingCall/teachingCallStatus/services/teachingCallStatusActionCreators.js [hunk 25]
```diff
@@ -77,9 +89,132 @@ teachingCallApp.service('teachingCallStatusActionCreators', function (teachingCa
...
+			teachingCallStatusStateService.reduce({
+				type: CALCULATE_INSTRUCTORS_IN_CALL,
+				payload: {
+					teachingCallsByInstructorType: teachingCallsByInstructorType,
+					instructorsInCalls: instructorsInCalls
+				}
+			});
 		}
 	};
 });
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/services/teachingCallStatusActionCreators.js]: End files with a newline

File: app/teachingCall/teachingCallStatus/templates/ModalAddInstructors.html [hunk 26]
```diff
@@ -15,48 +15,24 @@ <h4 class="modal-title">Add Instructors</h4>
...
-							ng-if="areAllLecturersInvited()"
-							disabled class="instructor-toggle-btn btn btn-sm"
-							uib-tooltip="All Lecturer Instructors Added"
-							id="courses-pivot">
-								Add Lecturers
-							</label>
+							<span ng-repeat="instructorTypeId in instructorTypes.ids">
+								<span ng-if="areThereInstructorsOfType(instructorTypeId)">
+									<label
+									ng-if="!areAllInstructorsOfTypeInvited(instructorTypeId)"
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/templates/ModalAddInstructors.html]: Multi-line tags should not have attributes lined up with tag name. Fix this in the rest of the file as well.

File: app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html [hunk 27]
```diff
@@ -32,7 +34,7 @@
 						<i style="padding-right: 5px;" class="glyphicon glyphicon-bullhorn"></i>
 						Contact Instructors
 					</a>
-					<span ng-if="atLeastOneInstructorSelected() == false" uib-tooltip="Select at least one instructor to contact">
+					<span ng-if="!(atLeastOneInstructorSelected())" uib-tooltip="Select at least one instructor to contact">
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html]: Switch back to "== false" for readability.

File: app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html [hunk 28]
```diff
@@ -45,219 +47,94 @@
...
-							message="Are you sure you want to remove this instructor?
-								This will cancel any pending notifications."
-							yes="Remove" no="Cancel" placement="top">
-						</span>
-					</div>
-				</div>
-			</div>
+		<div ng-show="view.state && view.state.ui.instructorsInCalls == false"
+		     class="teaching-call-status__no-calls-splash">
+			There are no instructors in a teaching call.
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html]: There are no instructors called. Use the 'Add Instructor' button above.

File: app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html [hunk 29]
```diff
@@ -45,219 +47,94 @@
...
-								terms-blob="{{instructor.termsBlob}}"
-								show-unavail="{{instructor.showUnavailabilities}}"
-								message="{{instructor.message}}"
-								placement="right">
+		<div ng-show="view.state && view.state.ui.instructorsInCalls">
+			<!-- Repeating sections -->
+			<div class="instructor-section-container"
+			     ng-repeat="instructorTypeId in view.state.instructorTypes.ids"
+			     ng-show="view.state.calculations.teachingCallsByInstructorType[instructorTypeId].length > 0">
+				<div class="instructor-section" style="width: 100%;">
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html]: Remove the inline styles found in this file

File: app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html [hunk 30]
```diff
@@ -45,219 +47,94 @@
...
+						<div class="column-last cell">
+							{{ instructor.lastContactedAt }}
+						</div>
+						<div class="column-next cell">
+							{{ instructor.nextContactAt }}
+						</div>
+						<div class="column-due cell">{{ instructor.dueDate }}</div>
+						<div class="column-hover-ui">
+							<span style="color: rgba(180, 20, 20, 0.8); font-size: 20px;"
+								class="remove-instructor-ui entypo-minus-circled text-primary clickable hovrable"
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html]: Line up attributes in multi-line tags

File: app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html [hunk 31]
```diff
@@ -45,219 +47,94 @@
...
-								This will cancel any pending notifications."
-							yes="Remove" no="Cancel" placement="top">
-						</span>
 					</div>
 				</div>
 			</div>
 		</div>
-
 	</div><!-- view.state check mask -->
 </div>
```
Comment (cthielen) [inline app/teachingCall/teachingCallStatus/templates/TeachingCallStatus.html]: End files with a newline

File: app/workgroup/controllers/UserCtrl.js [hunk 32]
```diff
@@ -7,53 +7,5 @@
...
-				workgroupActionCreators.createUser($scope.workgroupId, $scope.view.state.users.newUser);
-			};
-
-			$scope.removeUserFromWorkgroup = function(userId) {
-				var user = $scope.view.state.users.list[userId];
-				workgroupActionCreators.removeUserFromWorkgroup($scope.workgroupId, user);
-			};
-
+			// Intentionall Blank
 	}]);
```
Comment (cthielen) [inline app/workgroup/controllers/UserCtrl.js]: End files with a newline

File: app/workgroup/controllers/WorkgroupCtrl.js [hunk 33]
```diff
@@ -16,6 +16,7 @@ workgroupApp.controller('WorkgroupCtrl', ['$scope', '$rootScope', '$routeParams'
 
 		$rootScope.$on('workgroupStateChanged', function (event, data) {
 			$scope.view.state = data;
+			console.log(data);
```
Comment (cthielen) [inline app/workgroup/controllers/WorkgroupCtrl.js]: Remove debug code

File: app/workgroup/css/workgroup.css [hunk 34]
```diff
@@ -247,4 +247,9 @@ table th {
 
 .repeated-table-header {
 	margin-top: 0px;
+}
+
+.people-and-roles-nav-tab .tab-content {
+	border-left: 1px solid #ebebeb;
+	border-right: 1px solid #ebebeb;
 }
```
Comment (cthielen) [inline app/workgroup/css/workgroup.css]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/peopleAndRoles.html [hunk 35]
```diff
@@ -0,0 +1,13 @@
...
+	          tab-overrides="ui.roles.tabOverrides"
+	          select-tab="setRoleTab"
+	          slim="true">
+		<roles-table user-roles="calculatedUserRoles"
+		             active-role-id="ui.roles.activeRoleId"
+		             ui="ui"
+		             users="users">
+		</roles-table>
+	</ipa-tabs>
+</div>
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/peopleAndRoles.html]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/peopleAndRoles.js [hunk 36]
```diff
@@ -0,0 +1,17 @@
...
+			users: '<',
+			calculatedUserRoles: '<'
+		},
+		link: function(scope, element, attrs) {
+			scope.setRoleTab = function(tabName) {
+				workgroupActionCreators.setRoleTab(tabName);
+			};
+		}
+	};
+});
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/peopleAndRoles.js]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/instructorTypeSelector/instructorTypeSelector.html [hunk 37]
```diff
@@ -0,0 +1,7 @@
+<div class="instructor-type-selector">
+	<ipa-dropdown button-text="userRole.instructorTypeDescription || 'Assign instructor type'"
+	              items="instructorTypes"
+	              style="'minimal'"
+	              select-item="setInstructorType(item)">
+	</ipa-dropdown>
+</div>
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/instructorTypeSelector/instructorTypeSelector.html]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/instructorTypeSelector/instructorTypeSelector.js [hunk 38]
```diff
@@ -0,0 +1,16 @@
...
+			instructorTypes: '<',
+			userRole: '<'
+		},
+		link: function(scope, element, attrs) {
+			scope.setInstructorType = function(instructorType) {
+				workgroupActionCreators.setInstructorType(instructorType, scope.userRole);
+			};
+		}
+	};
+});
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/instructorTypeSelector/instructorTypeSelector.js]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.css [hunk 39]
```diff
@@ -0,0 +1,75 @@
...
+	align-items: center;
+}
+
+.roles-table__no-roles {
+	display: flex;
+	justify-content: center;
+	align-items: center;
+	padding: 10px;
+	color: #aaabae;
+}
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.css]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html [hunk 40]
```diff
@@ -0,0 +1,113 @@
...
+				</span>
+			</div>
+			<!-- /input-group -->
+		</div>
+		<!-- /.col-lg-6 -->
+	</div>
+
+	<table class="roles-table__table">
+		<colgroup>
+			<col span="1" style="width: 25%;">
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html]: Isn't span="1" and width=25% the default with 4 cols?

File: app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html [hunk 41]
```diff
@@ -0,0 +1,113 @@
...
+				<td class="roles-table__table-cell">
+					{{ userRole.userEmail }}
+				</td>
+
+				<!-- Dropdown for instructors -->
+				<td class="roles-table__table-cell">
+					<div class="roles-table__amount-container">
+						<!-- -1 is the category for 'student-ish' roles -->
+						<student-role-selector ng-show="activeRoleId == -1"
+																	 student-roles="ui.studentRoles"
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html]: Fix indentation

File: app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html [hunk 42]
```diff
@@ -0,0 +1,113 @@
...
+					</div>
+				</td>
+			</tr>
+
+			<tr ng-show="ui.roleTotals[activeRoleId] == 0">
+				<td colspan="4">
+				<div class="roles-table__no-roles">
+					No users
+				</div>
+			</td>
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html]: Fix indentation

File: app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html [hunk 43]
```diff
@@ -0,0 +1,113 @@
...
+			</tr>
+		</tbody>
+	</table>
+</div>
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.html]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.js [hunk 44]
```diff
@@ -0,0 +1,52 @@
...
+			};
+
+			scope.addUserToWorkgroup = function() {
+				scope.users.newUser;
+				workgroupActionCreators.createUser(scope.ui.workgroupId, scope.users.newUser, scope.activeRoleId);
+				scope.clearUserSearch();
+			};
+		}
+	};
+});
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/rolesTable.js]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/studentRoleSelector/studentRoleSelector.html [hunk 45]
```diff
@@ -0,0 +1,7 @@
+<div class="student-role-selector">
+	<ipa-dropdown button-text="userRole.description"
+	              items="studentRoles"
+	              style="'minimal'"
+	              select-item="updateStudentRole(item)">
+	</ipa-dropdown>
+</div>
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/studentRoleSelector/studentRoleSelector.html]: End files with a newline

File: app/workgroup/directives/peopleAndRoles/rolesTable/studentRoleSelector/studentRoleSelector.js [hunk 46]
```diff
@@ -0,0 +1,17 @@
...
+			userRole: '<'
+		},
+		link: function(scope, element, attrs) {
+			scope.updateStudentRole = function(role) {
+				scope.userRole.roleId = role.id;
+				workgroupActionCreators.updateStudentRole(scope.userRole);
+			};
+		}
+	};
+});
```
Comment (cthielen) [inline app/workgroup/directives/peopleAndRoles/rolesTable/studentRoleSelector/studentRoleSelector.js]: End files with a newline

File: app/workgroup/services/workgroupActionCreators.js [hunk 47]
```diff
@@ -142,41 +192,186 @@ workgroupApp.service('workgroupActionCreators', function (workgroupStateService,
...
+					activeRoleTab: tabName,
+					activeRoleTabDisplay: tabName == "Presence" ? "Unassigned" : tabName,
+					activeRoleId: this._getRoleIdFromTabName(tabName)
+				}
 			});
+		},
+		// Will return the roleId, or -1 for student (as it does not match a specific roleId, but will be expected to match 11,12 or 13 for masters, phd, instructional support)
+		_getRoleIdFromTabName: function(tabName) {
+			tabNameRoleIds = {
+				"Academic Planner": 2,
```
Comment (cthielen) [inline app/workgroup/services/workgroupActionCreators.js]: Perhaps these should be constants in a shared file?

File: app/workgroup/services/workgroupStateService.js [hunk 48]
```diff
@@ -155,35 +177,98 @@ workgroupApp.service('workgroupStateService', function ($rootScope, Role, Tag, L
...
+							allTabs: [
+								"Academic Planner",
+								"Instructor",
+								"Student",
+								"Presence"],
+						},
+						instructorTypes: [],
+						roleTotals: {},
+						studentRoles: [
+							{id: 11, description: "Instructional Support", name: "instructionalSupport"},
```
Comment (cthielen) [inline app/workgroup/services/workgroupStateService.js]: Remove magic numbers

File: app/budget/budgetCtrl.js (hunk 3 above)
Comment (ltwheeler) [reply app/budget/budgetCtrl.js]: The variable budgetConfigStyles will be handed to the generic ipa-modal component and used in an ng-style.

File: app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js (hunk 21 above)
Comment (ltwheeler) [reply app/teachingCall/teachingCallStatus/controllers/ModalAddInstructorsCtrl.js]: I like instructorTypeUsed() is better, I puzzled over areThereInstructorsOfType() but couldn't think of anything.

File: app/assignment/css/assignments.css [hunk 49]
```diff
@@ -961,5 +961,7 @@ li.active-term .dropdown-term-row {
 }
 
 .non-course-assignment {
-	 background-color: #6bc9f0;
+	border-color: rgb(199, 199, 199);
+	background-color: #f0f0f0;
+	color: rgb(123, 123, 123);
 }
```
Comment (cthielen) [inline app/assignment/css/assignments.css]: Files should end with newlines

File: app/budget/budgetApp.js [hunk 50]
```diff
@@ -59,6 +65,6 @@ var TOGGLE_FILTER_LINE_ITEM_SHOW_HIDDEN = "TOGGLE_FILTER_LINE_ITEM_SHOW_HIDDEN";
 var CALCULATE_SCENARIO_TERMS = "CALCULATE_SCENARIO_TERMS";
 var CALCULATE_SECTION_GROUPS = "CALCULATE_SECTION_GROUPS";
 var CALCULATE_TOTAL_COST = "CALCULATE_TOTAL_COST";
-var CALCULATE_INSTRUCTOR_TYPES = "CALCULATE_INSTRUCTOR_TYPES";
+var CALCULATE_INSTRUCTOR_TYPE_COSTS = "CALCULATE_INSTRUCTOR_TYPE_COSTS";
 var CALCULATE_INSTRUCTORS = "CALCULATE_INSTRUCTORS";
 var CALCULATE_LINE_ITEMS = "CALCULATE_LINE_ITEMS";
```
Comment (cthielen) [inline app/budget/budgetApp.js]: Files should end with newlines

File: app/budget/directives/modals/budgetConfig/groupCostConfig/groupCostConfig.html [hunk 51]
```diff
@@ -1,79 +1,50 @@
...
+					<ipa-input value="instructorTypeCost.cost"
+						place-holder="'Enter cost'"
+						mode="'currency'"
+						on-update="updateInstructorTypeCost(instructorTypeCost)">
+					</ipa-input>
+				</td>
+			</tr>
+		</tbody>
+	</table>
 </div>
```
Comment (cthielen) [inline app/budget/directives/modals/budgetConfig/groupCostConfig/groupCostConfig.html]: Files should end with newlines

File: app/budget/services/actions/budgetCalculations.js (hunk 13 above)
Comment (cthielen) [reply app/budget/services/actions/budgetCalculations.js]: Please address this

File: app/budget/services/actions/budgetCalculations.js (hunk 14 above)
Comment (cthielen) [reply app/budget/services/actions/budgetCalculations.js]: Please address this

File: app/budget/services/actions/budgetCalculations.js [hunk 52]
```diff
@@ -427,55 +505,164 @@ budgetApp.service('budgetCalculations', function ($rootScope, $window, budgetSer
...
+
+				if (teachingAssignment.instructorId == instructor.id) {
+					instructorType = instructorTypes.list[teachingAssignment.instructorTypeId];
+				}
+			});
+
+			return instructorType;
 		}
 	};
 });
```
Comment (cthielen) [inline app/budget/services/actions/budgetCalculations.js]: Files should end with newlines

File: app/scheduling/css/scheduling.css [hunk 53]
```diff
@@ -460,4 +460,9 @@ i.selected-section-group-remove {
 
 .filter-course--tag {
 	color: white;
+}
+
+.courses__metadata-container {
+	padding: 3px;
+	margin-left: 3px;
 }
```
Comment (cthielen) [inline app/scheduling/css/scheduling.css]: Files should end with newlines

File: app/shared/sharedApp.js [hunk 54]
```diff
@@ -127,6 +127,11 @@ sharedApp
...
-
 		}]
-	);
+	)
+	.constant('Roles', {
+		reviewer: 10,
+		instructor: 15,
+		academicPlanner: 2,
+		presence: 9
+	});
```
Comment (cthielen) [inline app/shared/sharedApp.js]: Files should end with newlines

File: app/teachingCall/css/teaching-call-status.css [hunk 55]
```diff
@@ -209,4 +219,61 @@
...
+}
+
+.teaching-call-status__instructor-details {
+	flex: 1 0 auto;
+	padding-right: 10px;
+}
+
+.teaching-call-status__instructor-tooltip {
+	padding-left:10px;
 }
```
Comment (cthielen) [inline app/teachingCall/css/teaching-call-status.css]: Files should end with newlines

File: app/teachingCall/teachingCallStatus/services/teachingCallStatusActionCreators.js (hunk 24 above)
Comment (cthielen) [reply app/teachingCall/teachingCallStatus/services/teachingCallStatusActionCreators.js]: Please address this

**Pull request merged**